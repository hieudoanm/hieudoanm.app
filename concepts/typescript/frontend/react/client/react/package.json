{
  "name": "@react/react",
  "version": "0.0.1",
  "description": "React - React",
  "main": "index.js",
  "repository": "git@github.com:hieudoanm/hieudoanm.git",
  "author": "hieudoanm <hieumdoan@gmail.com>",
  "license": "GPL-3.0",
  "private": true,
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build",
    "preview": "vite preview",
    "test": "jest",
    "lint": "eslint . --fix --ext .ts,.tsx"
  },
  "dependencies": {
    "@emotion/react": "11.11.3",
    "@emotion/styled": "11.11.0",
    "@fontsource/roboto": "5.0.8",
    "@mui/icons-material": "5.15.7",
    "@mui/material": "5.15.7",
    "@types/node": "20.11.13",
    "apexcharts": "3.45.2",
    "pino": "8.17.2",
    "react": "18.2.0",
    "react-apexcharts": "1.4.1",
    "react-dom": "18.2.0",
    "react-router-dom": "6.21.3"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "6.4.1",
    "@testing-library/react": "14.2.0",
    "@types/jest": "29.5.11",
    "@types/react": "18.2.48",
    "@types/react-dom": "18.2.18",
    "@typescript-eslint/eslint-plugin": "6.20.0",
    "@typescript-eslint/parser": "6.20.0",
    "@vitejs/plugin-react-swc": "3.6.0",
    "autoprefixer": "10.4.17",
    "eslint": "8.56.0",
    "eslint-plugin-react": "7.33.2",
    "jest": "29.7.0",
    "jest-environment-jsdom": "29.7.0",
    "kill-port": "2.0.1",
    "postcss": "8.4.33",
    "sass": "1.70.0",
    "tailwindcss": "3.4.1",
    "ts-jest": "29.1.2",
    "ts-node": "10.9.2",
    "typescript": "5.3.3",
    "vite": "5.0.12"
  },
  "keywords": [
    "react"
  ]
}
