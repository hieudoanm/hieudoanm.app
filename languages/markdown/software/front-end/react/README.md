---
title: 'React'
date: '2021-12-12'
author: 'Hieu Doan'
description: 'React'
---

### Class Component vs. Functional Component

- Difference syntax
- Before the introduction of hooks, functional components were stateless. However, with React 16.8, you can implement states with the useState hook to create a stateful component (just like the class component).
- Usage
  - Functional components with hooks are concise and more straightforward to code with. They perform exactly as the class component; this implies no difference between the two other than syntax.
  - By using just functional components in your project, you drastically eliminate the need to refactor the class component into a functional component when it grows.
  - Since classes confuse both people and machines, most especially the this keyword, you don't have to worry about this anymore in functional components.
  - No need for unnecessary method binding like we always do in the class component.
  - Sharing stateful logic between components is tedious in a class-based approach.

### `import`

- After compiling, the code we actually get is `return React.createElement`
