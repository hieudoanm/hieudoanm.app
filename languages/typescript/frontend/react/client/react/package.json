{
  "name": "@react/react",
  "version": "0.0.1",
  "description": "React - React",
  "main": "index.js",
  "repository": "git@github.com:houseoftypescript/web-react-template.git",
  "author": "hieudoanm <hieumdoan@gmail.com>",
  "license": "GPL-3.0",
  "private": true,
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build",
    "preview": "vite preview",
    "test": "jest",
    "lint": "eslint . --fix --ext .ts,.tsx"
  },
  "dependencies": {
    "@emotion/react": "11.11.1",
    "@emotion/styled": "11.11.0",
    "@fontsource/roboto": "5.0.8",
    "@mui/icons-material": "5.14.18",
    "@mui/material": "5.14.18",
    "@types/node": "20.9.1",
    "apexcharts": "3.44.0",
    "pino": "8.16.2",
    "react": "18.2.0",
    "react-apexcharts": "1.4.1",
    "react-dom": "18.2.0",
    "react-router-dom": "6.19.0"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "6.1.4",
    "@testing-library/react": "14.1.2",
    "@types/jest": "29.5.8",
    "@types/react": "18.2.37",
    "@types/react-dom": "18.2.15",
    "@typescript-eslint/eslint-plugin": "6.11.0",
    "@typescript-eslint/parser": "6.11.0",
    "@vitejs/plugin-react-swc": "3.5.0",
    "autoprefixer": "10.4.16",
    "eslint": "8.54.0",
    "eslint-plugin-react": "7.33.2",
    "jest": "29.7.0",
    "jest-environment-jsdom": "29.7.0",
    "kill-port": "2.0.1",
    "postcss": "8.4.31",
    "sass": "1.69.5",
    "tailwindcss": "3.3.5",
    "ts-jest": "29.1.1",
    "ts-node": "10.9.1",
    "typescript": "5.2.2",
    "vite": "5.0.0"
  },
  "keywords": [
    "react"
  ]
}
