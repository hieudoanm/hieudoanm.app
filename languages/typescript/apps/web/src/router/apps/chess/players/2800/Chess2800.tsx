import players from '@web/json/chess/players/2800.json';
import { Layout } from '@web/layout';
import Link from 'next/link';
import { FC } from 'react';

export const Chess2800: FC = () => {
  return (
    <Layout full nav>
      <div className='container mx-auto'>
        <div className='p-4 md:p-8'>
          <div className='rounded-xl border border-base-content'>
            <div className='p-2 md:p-4'>
              <div className='flex items-center justify-between'>
                <p>Players ({players.length})</p>
                <p>Peak Live Rating</p>
              </div>
            </div>
            {players.map(
              ({
                username,
                name,
                live,
              }: {
                username: string;
                name: string;
                live: number;
              }) => {
                return (
                  <div
                    key={username}
                    className='border-t border-base-content p-2 md:p-4'>
                    <div className='flex items-center justify-between'>
                      <Link href={`/apps/chess/players/2800/${username}`}>
                        {name}
                      </Link>
                      <p>{live}</p>
                    </div>
                  </div>
                );
              }
            )}
          </div>
        </div>
      </div>
    </Layout>
  );
};
