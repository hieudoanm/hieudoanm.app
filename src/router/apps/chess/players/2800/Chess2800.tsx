import players from '@web/json/chess/players/2800.json';
import { AppLayout } from '@web/layout/AppLayout';
import Link from 'next/link';
import { FC } from 'react';

export const Chess2800: FC = () => {
  return (
    <AppLayout full nav>
      <div className='container mx-auto'>
        <div className='p-4 md:p-8'>
          <div className='border-base-content rounded-xl border'>
            <table className='table'>
              <caption className='py-2'>2800 Clubs ({players.length})</caption>
              <thead>
                <tr className='border-base-content border-t'>
                  <th>Username</th>
                  <th>Name </th>
                  <th align='right'>Peak</th>
                </tr>
              </thead>
              <tbody>
                {players.map(
                  ({
                    username,
                    name,
                    live,
                  }: {
                    username: string;
                    name: string;
                    live: number;
                  }) => {
                    return (
                      <tr
                        key={username}
                        className='border-base-content border-t'>
                        <td>
                          <Link href={`/apps/chess/players/2800/${username}`}>
                            {username}
                          </Link>
                        </td>
                        <td>{name}</td>
                        <td align='right'>{live}</td>
                      </tr>
                    );
                  }
                )}
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </AppLayout>
  );
};
