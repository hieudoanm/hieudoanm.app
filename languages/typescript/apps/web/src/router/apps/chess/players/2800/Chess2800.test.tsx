import { render } from '@testing-library/react';
import { Chess2800 } from './Chess2800';

describe('Chess2800', () => {
  test('render default', () => {
    const wrapper = render(<Chess2800 />);
    expect(wrapper.container).toMatchSnapshot();
  });
});
