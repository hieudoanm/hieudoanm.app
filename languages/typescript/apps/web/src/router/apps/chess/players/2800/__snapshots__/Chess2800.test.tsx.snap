// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chess2800 render default 1`] = `
<div>
  <div
    class="flex flex-col h-screen overflow-hidden"
  >
    <nav>
      <div
        class=""
      >
        <div
          class="container mx-auto"
        >
          <div
            class="px-4 pt-4 md:px-8 md:pt-8 pb-0 md:pb-0"
          >
            <div
              class="flex items-center justify-between"
            >
              <a
                href="/"
              >
                <span
                  class="text-xl"
                >
                  <svg
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M3 19a2 2 0 0 0 2 2c2 0 2 -4 3 -9s1 -9 3 -9a2 2 0 0 1 2 2"
                    />
                    <path
                      d="M5 12h6"
                    />
                    <path
                      d="M15 12l6 6"
                    />
                    <path
                      d="M15 18l6 -6"
                    />
                  </svg>
                </span>
              </a>
              <a
                href="/apps"
              >
                <svg
                  class="hover:animate-spin"
                  fill="currentColor"
                  height="24"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M255.9 120.9l9.1-15.7c5.6-9.8 18.1-13.1 27.9-7.5 9.8 5.6 13.1 18.1 7.5 27.9l-87.5 151.5h63.3c20.5 0 32 24.1 23.1 40.8H113.8c-11.3 0-20.4-9.1-20.4-20.4 0-11.3 9.1-20.4 20.4-20.4h52l66.6-115.4-20.8-36.1c-5.6-9.8-2.3-22.2 7.5-27.9 9.8-5.6 22.2-2.3 27.9 7.5l8.9 15.7zm-78.7 218l-19.6 34c-5.6 9.8-18.1 13.1-27.9 7.5-9.8-5.6-13.1-18.1-7.5-27.9l14.6-25.2c16.4-5.1 29.8-1.2 40.4 11.6zm168.9-61.7h53.1c11.3 0 20.4 9.1 20.4 20.4 0 11.3-9.1 20.4-20.4 20.4h-29.5l19.9 34.5c5.6 9.8 2.3 22.2-7.5 27.9-9.8 5.6-22.2 2.3-27.9-7.5-33.5-58.1-58.7-101.6-75.4-130.6-17.1-29.5-4.9-59.1 7.2-69.1 13.4 23 33.4 57.7 60.1 104zM256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm216 248c0 118.7-96.1 216-216 216-118.7 0-216-96.1-216-216 0-118.7 96.1-216 216-216 118.7 0 216 96.1 216 216z"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="grow overflow-auto"
    >
      <div
        class="container mx-auto"
      >
        <div
          class="p-4 md:p-8"
        >
          <div
            class="rounded-xl border border-base-content"
          >
            <table
              class="table"
            >
              <caption
                class="py-2"
              >
                2800 Clubs (
                15
                )
              </caption>
              <thead>
                <tr
                  class="border-t border-base-content"
                >
                  <th>
                    Username
                  </th>
                  <th>
                    Name 
                  </th>
                  <th
                    align="right"
                  >
                    Peak
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/magnuscarlsen"
                    >
                      magnuscarlsen
                    </a>
                  </td>
                  <td>
                    Magnus Carlsen
                  </td>
                  <td
                    align="right"
                  >
                    2889.2
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/garrykasparov"
                    >
                      garrykasparov
                    </a>
                  </td>
                  <td>
                    Garry Kasparov
                  </td>
                  <td
                    align="right"
                  >
                    2856.7
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/fabianocaruana"
                    >
                      fabianocaruana
                    </a>
                  </td>
                  <td>
                    Fabiano Caruana
                  </td>
                  <td
                    align="right"
                  >
                    2851.3
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/levonaronian"
                    >
                      levonaronian
                    </a>
                  </td>
                  <td>
                    Levon Aronian
                  </td>
                  <td
                    align="right"
                  >
                    2835.5
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/gmwso"
                    >
                      gmwso
                    </a>
                  </td>
                  <td>
                    Wesley So
                  </td>
                  <td
                    align="right"
                  >
                    2824.5
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/azerichess"
                    >
                      azerichess
                    </a>
                  </td>
                  <td>
                    Shakhriyar Mamedyarov
                  </td>
                  <td
                    align="right"
                  >
                    2826.2
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/lyonbeast"
                    >
                      lyonbeast
                    </a>
                  </td>
                  <td>
                    Maxime Vachier-Lagrave
                  </td>
                  <td
                    align="right"
                  >
                    2819.3
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/thevish"
                    >
                      thevish
                    </a>
                  </td>
                  <td>
                    Viswanathan Anand
                  </td>
                  <td
                    align="right"
                  >
                    2820.7
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/vladimirkramnik"
                    >
                      vladimirkramnik
                    </a>
                  </td>
                  <td>
                    Vladimir Kramnik
                  </td>
                  <td
                    align="right"
                  >
                    2817.8
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/veselintopalov359"
                    >
                      veselintopalov359
                    </a>
                  </td>
                  <td>
                    Veselin Topalov
                  </td>
                  <td
                    align="right"
                  >
                    2826.5
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/hikaru"
                    >
                      hikaru
                    </a>
                  </td>
                  <td>
                    Hikaru Nakamura
                  </td>
                  <td
                    align="right"
                  >
                    2819
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/chefshouse"
                    >
                      chefshouse
                    </a>
                  </td>
                  <td>
                    Ding Liren
                  </td>
                  <td
                    align="right"
                  >
                    2816.9
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/grischuk"
                    >
                      grischuk
                    </a>
                  </td>
                  <td>
                    Alexander Grischuk
                  </td>
                  <td
                    align="right"
                  >
                    2813.6
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/firouzja2003"
                    >
                      firouzja2003
                    </a>
                  </td>
                  <td>
                    Alireza Firouzja
                  </td>
                  <td
                    align="right"
                  >
                    2804
                  </td>
                </tr>
                <tr
                  class="border-t border-base-content"
                >
                  <td>
                    <a
                      href="/apps/chess/players/2800/anishgiri"
                    >
                      anishgiri
                    </a>
                  </td>
                  <td>
                    Anish Giri
                  </td>
                  <td
                    align="right"
                  >
                    2802.8
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
